iree/compiler/Dialect/Stream/Conversion/FlowToStream/ConvertFlowToStream.cpp

```diff
@@ -0,0 +1,472 @@
+                                           op->getAttrDictionary());
+
+    // Query all operand sizes.
+    SmallVector<Value> operandSizes;
+    for (auto oldNewOperand : llvm::zip(op.operands(), operands.operands())) {
+      auto oldOperand = std::get<0>(oldNewOperand);
+      auto newOperand = std::get<1>(oldNewOperand);
+      if (oldOperand.getType().isa<ShapedType>()) {
+        auto operandDynamicDims = Shape::buildOrFindDynamicDimsForValue(
+            op.getLoc(), oldOperand, rewriter);
+        operandSizes.push_back(querySizeOf(op.getLoc(), newOperand, oldOperand,
+                                           operandDynamicDims, rewriter));
+      }
+    }
+
+    // Construct result sizes or reuse tied operand sizes from above.
+    SmallVector<Value> resultSizes;
+    SmallVector<Type> resultTypes;
+    auto unknownType = IREE::Stream::ResourceType::get(getContext());
+    auto tiedOperandBase = op.getTiedOperandsIndexAndLength().first;
+    for (auto result : llvm::enumerate(op.results())) {
+      auto oldResultType = result.value().getType();
+      if (!oldResultType.isa<ShapedType>()) {
+        resultTypes.push_back(getTypeConverter()->convertType(oldResultType));
+        continue;
+      }
+      auto tiedOperand = op.getTiedResultOperandIndex(result.index());
+      if (tiedOperand.hasValue()) {
+        auto operandIndex = tiedOperand.getValue() - tiedOperandBase;
+        resultSizes.push_back(operandSizes[operandIndex]);
+        resultTypes.push_back(operands.operands()[operandIndex].getType());
+      } else {
+        auto resultDynamicDims = Shape::buildOrFindDynamicDimsForValue(
+            op.getLoc(), result.value(), rewriter);
+        resultSizes.push_back(buildResultSizeOf(op.getLoc(), result.value(),
+                                                resultDynamicDims, rewriter));
+        resultTypes.push_back(unknownType);
+      }
+    }
+
+    rewriter.replaceOpWithNewOp<IREE::Stream::AsyncDispatchOp>(
+        op, resultTypes, operands.workgroup_count(), operands.entry_point(),
+        operands.operands(), operandSizes, resultSizes,
+        operands.tied_operands(),
+        /*affinity=*/nullptr);
+    return success();
+  }
+};
+
+static SmallVector<Value> makeBindingDynamicDims(
+    Location loc, IREE::Flow::DispatchTensorType tensorType, BlockArgument arg,
+    OpBuilder &builder) {
+  if (tensorType.hasStaticShape()) return {};
+
+  // We can expect its first user to be a tie_shape op to associate
+  // concrete dimension values. Originally we have such information
+  // maintained in the flow ops handling dynamic tensors. But during
+  // flow executable outlining, such information is transfered to
+  // tie_shape ops.
+  //
+  // HACK: this is disgusting - we should carry this from the flow level in the
+  // right way such that we don't need to make this assumption.
+  auto tieShapeOp = cast<IREE::Flow::DispatchTieShapeOp>(*arg.user_begin());
```

Comment:
Fixed.

The DispatchTieShapeOp needs to die - until it does, such hackery is required :( But I can't change any of that so long as the old non-streams path still exists and removing this is high on my list (as well as removing the shape dialect) so I'll be doing that in a follow-up. Today the only thing that inserts these ops is the dispatch region outlining pass and so the only time this should fail is if that pass is broken - which would be if I changed something - so I'm fine riding with this until we can kill it all. We finally have a bot with asserts (right?) and it really is an internal error if we generate things wrong. I'd like people to be able to come in with outlined dispatch regions already, however the ergonomics here with these shape ops is not good for that anyway and I'd rather make it easier to use instead of adding user-facing failure paths that aren't actionable. *shakes fist* SHAAAPEESESSSSS